pcap-dpdk.c

```diff
@@ -0,0 +1,742 @@
+	struct pcap_dpdk * pd = p->priv;
+	int ret=0;
+	ret = install_bpf_program(p, fp); 
+	if (ret==0){
+		pd->filter_in_userland = 1;
+	}
+	return ret;
+}
+
+static void nic_stats_display(struct pcap_dpdk *pd)
+{
+	uint16_t portid = pd->portid;
+	struct rte_eth_stats stats;
+	rte_eth_stats_get(portid, &stats);
+	RTE_LOG(INFO,USER1, "portid:%d, RX-packets: %-10"PRIu64"  RX-errors:  %-10"PRIu64
+	       "  RX-bytes:  %-10"PRIu64"  RX-Imissed:  %-10"PRIu64"\n", portid, stats.ipackets, stats.ierrors,
+	       stats.ibytes,stats.imissed);
+	RTE_LOG(INFO,USER1, "portid:%d, RX-PPS: %-10"PRIu64" RX-Mbps: %.2lf\n", portid, pd->pps, pd->bps/1e6f );
+}
+
+static int pcap_dpdk_stats(pcap_t *p, struct pcap_stat *ps)
+{
+	struct pcap_dpdk *pd = p->priv;
+	calculate_timestamp(&(pd->ts_helper), &(pd->curr_ts));
+	rte_eth_stats_get(pd->portid,&(pd->curr_stats));
+
+	ps->ps_recv = pd->curr_stats.ipackets;
+	ps->ps_drop = pd->curr_stats.ierrors;
+	ps->ps_drop += pd->bpf_drop;
+	ps->ps_ifdrop = pd->curr_stats.imissed;
+	
+	uint64_t delta_pkt = pd->curr_stats.ipackets - pd->prev_stats.ipackets;
+	struct timeval delta_tm;
+	timersub(&(pd->curr_ts),&(pd->prev_ts), &delta_tm);
+	uint64_t delta_usec = delta_tm.tv_sec*1e6+delta_tm.tv_usec;
+	uint64_t delta_bit = (pd->curr_stats.ibytes-pd->prev_stats.ibytes)*8;
+	RTE_LOG(INFO, USER1, "delta_usec: %-10"PRIu64" delta_pkt: %-10"PRIu64" delta_bit: %-10"PRIu64"\n", delta_usec, delta_pkt, delta_bit);
+	pd->pps = (uint64_t)(delta_pkt*1e6f/delta_usec);
+	pd->bps = (uint64_t)(delta_bit*1e6f/delta_usec);
+	nic_stats_display(pd);
+	pd->prev_stats = pd->curr_stats;
+	pd->prev_ts = pd->curr_ts;
+	return 0;
+}
+
+static int pcap_dpdk_setnonblock(pcap_t *p, int fd _U_){
+	pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+		errno, "dpdk error: setnonblock not support");
+	return 0;
+}
+
+static int pcap_dpdk_getnonblock(pcap_t *p){
+	pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+		errno, "dpdk error: getnonblock not support");
+	return 0;
+}
+
+static int check_link_status(uint16_t portid, struct rte_eth_link *plink)
+{
+	uint8_t count = 0;
+	int is_port_up = 0;
+	int max_check_time = 50;
+	int check_interval = 100; // 100ms
+	for (count = 0; count <= max_check_time; count++) {
+		memset(plink, 0, sizeof(struct rte_eth_link));
+		rte_eth_link_get_nowait(portid, plink);
+		if (plink->link_status == ETH_LINK_UP)
+		{
+			is_port_up = 1;
+			break;
+		}else{
+			rte_delay_ms(check_interval);
+		}
+	}
+	return is_port_up;
+}
+
+// return portid by device name, otherwise return -1
+static uint16_t portid_by_device(char * device)
+{
+	uint16_t ret = -1;
+	int len = strlen(device);
+	int prefix_len = strlen(DPDK_PREFIX);
+	unsigned long ret_ul = 0L;
+
+	if (len<=prefix_len || strncmp(device, DPDK_PREFIX, prefix_len)) // check prefix dpdk:
+	{
+		return ret;
+	}
+	if (device[prefix_len]>='0' && device[prefix_len]<='9')
+	{ // is digital
+		ret_ul = strtoul(&(device[prefix_len]), NULL, 10);
```

Comment:
This should check whether `strtoul()` found a number.

That means that the second argument should point to a `char *` pointer and, when it returns, you should report an error if either 1) the pointer is equal to `&device[prefix_len])` (meaning the first character wasn't a digit) or 2) the pointer doesn't point to a `\0' (meaning that there's a non-digit in the string).

However, you should still do the test for the first character being a digit, because, for some reason, the argument to `strtoul()` can begin with a `-`, the "u" in "strtoul" standing for "unsigned" nonwithstanding.

Reply:
OK, I will check the rest char of device are all digital then to the strtoul.
The following bad cases will be rejected: 
dpdk:1e2
dpdk:1u